types/trigger.d.ts

```diff
@@ -43,34 +46,110 @@ export type Output = {
-  [key: string]: unknown;
-}
-
-export type TimelineTrigger = Trigger & { beforeSeconds: number };
+export type ResponseField = ResponseFunc<MatchesAny> | ResponseOutput<MatchesAny>;
 
 export type TriggerAutoConfig = {
   Output?: Output;
   Duration?: number;
   BeforeSeconds?: number;
   OutputStrings?: { [outputKey: string]: Lang };
 }
+
+export type MatchesAny = { [s in T]?: string } | undefined;
+export type TriggerField<Return> = TriggerFunc<MatchesAny, Return> | Return;
+
+// TODO: I am not sure that it is possible to type triggers such that we have some gigantic union
+// of Trigger<TetherParams> | Trigger<AbilityParams> | Trigger<AddedCombatantParams> | etc etc
+// and so the idea is that individual files would use a narrower type for Matches if they wanted to
+// and that's something we (maybe?) could enforce via tests rather than via TypeScript.  This
+// approach would not preclude some TypeScript gigabrain solution that can figure out how to do
+// the same thing at the TypeScript level.
+//
+// TODO: I think there are two types of TriggerSets.  There is the "cactbot + user defined triggers"
+// and also the "cactbot built-in only" triggers, which can be more strict.  For example,
+// * is `id` required, and `regex` for timeline triggers? (yes for built-in, no for user triggers)
+// * should we handle nonsense like `condition` returning non-booleans?
+// We could have one type that built-in triggers define themselves as more strictly and then a
+// second type that popup-text.js uses and needs to do more narrowing for.
+export type BaseTrigger = {
+  id?: string;
+  disabled?: boolean;
+  condition?: TriggerField<boolean | undefined>;
+  preRun?: TriggerField<void>;
+  delaySeconds?: TriggerField<number | undefined>;
+  durationSeconds?: TriggerField<number | undefined>;
+  suppressSeconds?: TriggerField<number | undefined>;
+  promise?: TriggerField<Promise<void> | undefined>;
+  sound?: TriggerField<string | undefined>;
+  soundVolume?: TriggerField<number | undefined>;
+  response?: ResponseField;
+  alarmText?: TriggerField<TriggerOutput>;
+  alertText?: TriggerField<TriggerOutput>;
+  infoText?: TriggerField<TriggerOutput>;
+  tts?: TriggerField<TriggerOutput>;
+  run?: TriggerField<void>;
+  outputStrings?: {
+    [key: string]: LocaleText;
+  };
+}
+
+export type NetRegexTrigger = BaseTrigger & {
+  netRegex: RegExp;
+  netRegexDe?: RegExp;
+  netRegexFr?: RegExp;
+  netRegexJa?: RegExp;
+  netRegexCn?: RegExp;
+  netRegexKo?: RegExp;
+}
+
+export type RegexTrigger = BaseTrigger & {
+  regex: RegExp;
+  regexDe?: RegExp;
+  regexFr?: RegExp;
+  regexJa?: RegExp;
+  regexCn?: RegExp;
+  regexKo?: RegExp;
+}
+
+export type TimelineTrigger = BaseTrigger & {
+  regex?: RegExp;
+  // TODO: can this also be a function?
+  beforeSeconds?: number;
+};
+
+export type Trigger = RegexTrigger | NetRegexTrigger;
+
+export type TimelineFunc = string | string[] | ((data: Data) => TimelineFunc);
+
+export type TriggerSet = {
```

Comment:
This should be a genetic type I think. We need to extend `Data` for a specific trigger set.

Reply:
I am not sure that's true.  Because `data` is reset to the base `data`, then every field in `TEAData` must be optional.  And, because all fields are optional, you can do what I did in `t1.ts`, where I have `ThisData extends Data` and use it locally inside of that function in place of `Data`.  In other words, I can imagine each trigger file having its own type that extends the base data type and using that as a parameter for Data locally.

Reply:
code outside the trigger file itself didn't need to know the extend type, they just need to know it's a `TriggerSet<Data>`.

Reply:
Oh sorry, I import `TEAData` because I have multiple trigger files for TEA, not because I'm using this type outside the trigger files.

Reply:
Oh oh oh, I see!  I get it.  Thanks, that's a good idea!

Reply:
Done in the latest patch, what a great idea!